---
title: "Investigating energy metabolism in *C. necator* using a barcoded transposon library"
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Michael Jahn"
output:
  html_notebook: 
    theme: cosmo
    toc: yes
    number_sections: true
---

# Description

This R notebook is a bioinformatics pipeline to analyze fitness data obtained from a barcoded transposon library in *Ralstonia eutropha* a.k.a. *Cupriavidus necator*. For background and details regarding the method, see [Wetmore at al., mBio, 2015](https://mbio.asm.org/content/6/3/e00306-15) and [Price et al., Nature, 2018](http://www.nature.com/articles/s41586-018-0124-0)).

One part of the data used here (growth on **fructose**, **succinate**, **formate**) was already published in a previous research article with focus on *C. necator*'s central carbon metabolism. The [preprint is available on BioRxiv](https://www.biorxiv.org/content/10.1101/2021.03.21.436304v2). The second part (autotrophic growth on **hydrogen**, anoxic growth with **nitrate respiration**) is new data that complements previous results. The focus of this pipeline is to study the energy metabolism of *C. necator* in diverse conditions, each of it challenging a different aspect/machinery of metabolism. This investigation tries to answer the following key questions:

1. Which genes are constituting the essential parts of the **hydrogenases and formate dehydrogenases**, enabling autotrophic lifestyle? Many things about the structure and function of these enzymes is already known, where do these results show differences?
2. Which genes are **mandatory as accessory proteins**, for example cofactor synthesis, insertion, and enzyme maturation?
3. Why does **KO of hydrogenases provides a growth benefit** in non-autotrophic conditions?
4. What are the essential parts of the **electron transport chain**? What complexes are used specifically for some growth conditions?
5. Is essentiality/fitness correlated with protein abundance? Is there a dosage effect, such that KO of a high abundant iso-enzyme has stronger fitness effect compared to a low-abundant?


# Libraries

Optionally install required libraries, particularly custom packages from github.

```{r, eval = FALSE}
devtools::install_github("m-jahn/lattice-tools")
```


```{r}
suppressPackageStartupMessages({
  library(lattice)
  library(latticeExtra)
  library(latticetools)
  library(tidyverse)
  library(dendextend)
  library(colorspace)
  library(stringi)
  library(vegan)
  library(tsne)
})
```


# Data import and processing

Import the main data tables with gene fitness data for all conditions. Tables were obtained by processing sequencing data with a custom [BarSeq pipeline](https://github.com/m-jahn/rebar). The 32 generation sequencing samples are removed due to low diversity, leading to overall lower average read count per gene.

```{r, message = FALSE}
load("../data/barseq/20201222_fru_fitness.Rdata")
df_fitness_frc <- fitness_gene %>% ungroup %>%
  filter(condition != "long pulse", time != 32) %>%
  mutate(ID = as.numeric(ID), substrate = "fructose",
    condition = str_remove(condition, "short "))

load("../data/barseq/20210407_suc_for_fitness.Rdata")
df_fitness_suc <- fitness_gene %>% ungroup %>%
  separate(condition, sep = "_", into = c("substrate", "condition"))

load("../data/barseq/20210624_H2_NO3_fitness.Rdata")
df_fitness_no3 <- fitness_gene %>% ungroup %>%
  filter(condition %in% c("NO3")) %>%
  mutate(substrate = "nitrate", condition = "continuous")

load("../data/barseq/20210928_H2_fitness.Rdata")
df_fitness_h2 <- fitness_gene %>% ungroup %>%
  mutate(substrate = "hydrogen", condition = "pulse")

# merge fitness tables
df_fitness <- bind_rows(df_fitness_frc, df_fitness_suc, df_fitness_no3, df_fitness_h2) %>%
  rename(locus_tag = locusId)
rm("df_fitness_frc", "df_fitness_suc", "df_fitness_h2", "df_fitness_no3", "fitness_gene")

# import genome annotation
df_ref <- read_csv("../data/ref/Ralstonia_H16_genome_annotation.csv",
  col_types = cols()) %>%
  filter(!duplicated(locus_tag)) %>%
  mutate(eggNOG_name = if_else(is.na(eggNOG_name), gene_name, eggNOG_name))

# define standard colors
stdcol <- custom.colorblind()$superpose.line$col
```

# Gene fitness analysis

## Depletion over time (generations)

We can plot log2 FC or normalized gene fitness over generations. For this type of overview it is best to summarize individual replicates (4x) to the mean or median, per time point and condition. We also add genome annotation to the summary table. The plots shows that depletion of some strains is so strong already at 8 generations that fitness/log2 FC could not be quantified for 16 generations due to missing read counts. It is best to focus on 8 generations as because it provides a more complete picture.

```{r, messages = FALSE}
df_fitness_summary <- df_fitness %>%
  group_by(locus_tag, scaffold, time, condition, substrate, strains_per_gene) %>%
  summarize(
    norm_gene_fitness_median = median(norm_gene_fitness, na.rm = TRUE),
    log2FC_median = median(log2FC, na.rm = TRUE),
    tstat_median = median(t, na.rm = TRUE)
  ) %>%
  left_join(df_ref) %>% ungroup %>%
  mutate(cond_short = if_else(condition == "pulse", "P", "C")) %>%
  mutate(cond_short = paste0(substrate, " - ", cond_short) %>% factor(., unique(.)[c(2,6,4,8,1,5,7,3)]))

levels(df_fitness_summary$cond_short)
```


```{r, fig.width = 6.5, fig.height = 8}
plot_hist_log2FC <- df_fitness_summary %>%
  group_by(condition, substrate) %>%
  slice(1:1000) %>%
  
  xyplot(log2FC_median ~ time | substrate * condition, .,
    groups = locus_tag, as.table = TRUE,
    col = stdcol[1], alpha = 0.2, layout = c(5, 2),
    xlab = "", ylab = expression("log"[2]*" FC"),
    par.settings = custom.colorblind(), type = c("l"),
    between = list(x = 0.5, y = 0.5),
    scales = list(alternating = FALSE), lwd = 2,
    panel = function(x, y, ...) {
      panel.grid(h = -1, v = -1, col = grey(0.9))
      panel.xyplot(x, y, ...)
    }
  )

plot_hist_normfg <- df_fitness_summary %>%
  group_by(condition, substrate) %>%
  slice(1:1000) %>%
  
  xyplot(norm_gene_fitness_median ~ time | substrate * condition, .,
    groups = locus_tag, as.table = TRUE,
    col = stdcol[2], alpha = 0.2, layout = c(5, 2),
    xlab = "generations", ylab = "fitness",
    par.settings = custom.colorblind(), type = c("l"),
    between = list(x = 0.5, y = 0.5),
    scales = list(alternating = FALSE), lwd = 2,
    panel = function(x, y, ...) {
      panel.grid(h = -1, v = -1, col = grey(0.9))
      panel.xyplot(x, y, ...)
    }
  )

print(plot_hist_log2FC, position = c(0,0.47,1,1), more = TRUE)
print(plot_hist_normfg, position = c(0,0,1,0.53))
```

```{r, echo = FALSE}
png("../figures/figure_depletion_over_time.png", width = 900, height = 1200, res = 150)
print(plot_hist_log2FC, position = c(0,0.47,1,1), more = TRUE)
print(plot_hist_normfg, position = c(0,0,1,0.53))
dev.off()
```


## Comparing gene fitness between conditions

Similarly to Figure 2 in [Wetmore et al, mBIO, 2015](https://mbio.asm.org/content/6/3/e00306-15), we can investigate condition-dependent gene fitness by comparing conditions and substrates one-on-one. The correlation between substrates and growth regimes (continuous, pulsed) is quite different. The strongest correlation is present in the pulsed vs continuous regime for each substrate (R = 0.64 to 0.87), and within the pulsed conditions the comparison between substrates (R = 0.62 to 0.72).

```{r, fig.width = 8, fig.height = 8, message = FALSE}
df_fitness_comp <- df_fitness_summary %>% filter(time == 8) %>%
  group_by(locus_tag) %>% mutate(tstat_median = min(tstat_median)) %>%
  select(locus_tag, cond_short, norm_gene_fitness_median, gene_name, COG_Process, tstat_median) %>%
  pivot_wider(names_from = cond_short, values_from = norm_gene_fitness_median) %>%
  filter(!is.na(locus_tag)) %>% ungroup

custom_splom(df_fitness_comp %>%
  select(matches(" - [CP]")), pch = 19, cex = 0.3, col = grey(0.4, 0.4))
```

```{r, echo = FALSE}
png("../figures/figure_splom_all_conditions.png", width = 1200, height = 1200, res = 150)
custom_splom(df_fitness_comp %>%
  select(matches(" - [CP]")), pch = 19, cex = 0.3, col = grey(0.4, 0.4)) %>% print
dev.off()
```

# Unsupervised clustering of genes by fitness

Most genes correlate in fitness value between conditions. That means, only some genes have a **condition-specific fitness effect**, i.e. increase or decrease fitness of the respective strain in one substrate or growth regime specifically. This section investigates specific genes and their functions that show such differential fitness between conditions. A simple comparison of one condition versus another will not be helpful as there two many possible combinations. To identify interesting sets of genes, cluster analysis can be performed. The first step is to generate a matrix from reformatted fitness data, and  cluster genes by similarity using a generic clustering algorithm such as `hclust(method = "ward.D")`.

```{r}
# generate colorpalette for heatmap
heat_cols <- diverging_hcl(n = 7, h = c(255, 12), c = c(50, 80), l = c(20, 97), power = c(1, 1.3))

# create a matrix from wide fitness data for plotting heatmap
mat_heatmap <- df_fitness_comp %>% ungroup %>%
  filter(if_any(.cols = matches(" - [CP]"), ~ !between(., -2, 2))) %>%
  select(matches("locus_tag| - [CP]")) %>%
  # filter out NA rows, and replace extreme values
  drop_na %>% mutate(across(matches(" - [CP]"), 
    function(x) {y = replace(x, x > 6, 6); replace(y, y < -6, -6)})) %>%
  column_to_rownames(var = "locus_tag") %>%
  as.matrix

# create cluster for reordering
mat_cluster <- mat_heatmap %>% dist %>% hclust(method = "ward.D2")
mat_heatmap <- mat_heatmap[order.dendrogram(as.dendrogram(mat_cluster)), c(1,5,2,6,4,8,3,7)]
```

The second step is to find the optimal number of clusters for subdividing the data set. Silhouette analysis can be used for this purpose. However, the result is not entirely clear, but suggests a number of 4 to 10 clusters would give good separation.

```{r, fig.width = 4, fig.height = 4}
#devtools::install_github("m-jahn/Rtools")
set.seed(123)
silhouette_result <- Rtools::silhouette_analysis(mat = mat_heatmap, n = 2:20)
print(silhouette_result$plot.summary)
n_clusters = 7
```


```{r, fig.width = 10, fig.height = 4.0, warning = FALSE}
# define a custom set of colors for clusters
custom_colors <- colorRampPalette(custom.colorblind()$superpose.line$col[c(2,3,1,5)])(n_clusters)

# plot heatmap
plot_heatmap <- levelplot(mat_heatmap,
  par.settings = custom.colorblind(),
  col.regions = colorRampPalette(heat_cols)(16),
  at = seq(-6, 6, 1), aspect = "fill",
  xlab = "", ylab = "", scales = list(x = list(draw = FALSE)),
  panel = function(x, y, z, ...) {
    panel.levelplot(x, y, z, ...)
    panel.abline(h = 1:5+0.5, col = "white", lwd = 1.5)
  }
)

plot_cluster_dend <- mat_cluster %>% as.dendrogram %>%
  set("branches_k_col", custom_colors, k = n_clusters) %>%
  set("branches_lwd", 0.5) %>%
  as.ggdend %>%
  ggplot(labels = FALSE)

gridExtra::grid.arrange(
  # coords for unit: top, right, bottom, left
  plot_cluster_dend + 
    theme(plot.margin = unit(c(0.1, 0.05, -0.26, 0.085),"npc")),
  plot_heatmap,
  nrow = 2
)
```


The heat map clusters genes together that show similar expression over different conditions. Alternatively, we can use dimensionality reduction to identify clusters of genes that are similar, e.g. through **PCA**, **NMDS**, or **t-SNE**. 


```{r, fig.width = 8, fig.height = 4}
# set a seed to obtain same pattern for stochastic methods
set.seed(123)

# run nMDS analysis
NMDS <- mat_heatmap %>% dist %>% metaMDS
df_nmds <- NMDS$points %>% as_tibble(rownames = "locus") %>%
  left_join(enframe(cutreeord(mat_cluster, k = n_clusters), "locus", "cluster"))

# run t-SNE analysis
SNE <- tsne(mat_heatmap %>% dist)
df_tsne <- SNE %>% setNames(c("x", "y")) %>% as_tibble %>%
  mutate(locus = rownames(mat_heatmap)) %>%
  left_join(enframe(cutreeord(mat_cluster, k = n_clusters), "locus", "cluster"))

plot_nmds <- df_nmds %>% 
  xyplot(MDS2 ~ MDS1, .,
    groups = cluster, col = custom_colors,
    par.settings = custom.colorblind(),
    aspect = 1, main = "nMDS",
    panel = function(x, y, ...) {
      panel.grid(h = -1, v = -1, col = grey(0.9))
      panel.xyplot(x, y, ...)
      panel.key(..., cex = 0.7, corner = c(0.02, 0.98))
    }
  )

plot_tsne <- df_tsne %>% 
  xyplot(V2 ~ V1, .,
    groups = cluster, col = custom_colors,
    par.settings = custom.colorblind(),
    aspect = 1,  main = "t-SNE",
    panel = function(x, y, ...) {
      panel.grid(h = -1, v = -1, col = grey(0.9))
      panel.xyplot(x, y, ...)
      panel.key(..., cex = 0.7, corner = c(0.02, 0.98))
    }
  )

print(plot_nmds, split = c(1,1,2,1), more = TRUE)
print(plot_tsne, split = c(2,1,2,1))
```


```{r, echo = FALSE}
svg(filename = "../figures/figure_nmds_tsne.svg", width = 8, height = 4)
print(plot_nmds, split = c(1,1,2,1), more = TRUE)
print(plot_tsne, split = c(2,1,2,1))
dev.off()
```

# Supervised analysis of energy metabaolism

Generalized function plot (small) heatmaps.

```{r}
# generalized function to plot (small) heatmaps
heatmap_fitness <- function(data, key = TRUE) {
  
  # make some small adjustments to optical appearance
  data %>%
  mutate(norm_gene_fitness_median = norm_gene_fitness_median %>%
    replace(., . < -6, -6) %>% replace(., . > 6, 6)) %>%

  # plot heatmap
  levelplot(norm_gene_fitness_median ~ factor(gene_name) * cond_short, .,
    par.settings = custom.colorblind(), colorkey = key,
    col.regions = colorRampPalette(heat_cols)(16),
    at = seq(-6, 6, 1), aspect = "iso",
    xlab = "", ylab = "", scales = list(cex = 0.7, x = list(rot = 90)),
    panel = function(x, y, z, ...) {
      panel.levelplot(x, y, z, ...)
      panel.abline(v = seq_along(unique(x))+0.5, 
        h = seq_along(unique(y))+0.5, col = "white", lwd = 2)
    }
  )
}
```

Generalized function to plot genes for different regions of the genome.

```{r}
genome_plot <- function(df, xlim = NULL, title = "") {
  theme = custom.colorblind(col = c(grey(0.7), grey(0.85)))
  theme$axis.line$col = grey(0.3)
  theme$axis.text$col = grey(0.3)
  theme$axis.text$cex = 0.7
  
  if (!is.null(xlim))
    xscale = list(limits = xlim)
  else 
    xscale = list()
  xyplot(end/1000 ~ start/1000, df,
    groups = strand, cex = 0.7, lwd = 1,
    par.settings = theme, strains = df$strains_per_gene,
    scales = list(y = list(draw = FALSE), x = xscale),
    ylim = c(-3,2), xlab = "", ylab = "",
    gene_strand = df[["strand"]],
    gene_name = df[["gene_name"]],
    panel = function(x, y, strains, ...) {
      panel.geneplot(x, y, arrows = TRUE, 
        tip = 0.1,  rot_labels = 0, ...)
      panel.text((x+y)/2, rep(0, length(x)), labels = strains, cex = 0.7)
      panel.text(mean(xlim), 1.5, labels = title, col = 1)
    }
  )
}
```

## Molybdenum cofactor biosynthesis

- modAB = molybdenum transporter
- moaACDE = molybdopterin backbone biosynthesis
- moeA = molybdopterin molybdenum transferase
- mobAB = molybdenum cofactor guanylyltransferase, 
- mog = molybdopterin biosynthesis protein (role unclear?)


```{r, fig.width = 7, fig.height = 4}
# plot heatmap
df_moco <- df_fitness_summary %>% filter(time == 8, grepl("mo[aedbg]", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "mo[aedbg]([A-Z0-9]+)?")) %>%
  ungroup %>% arrange(gene_name)
plot_moco_fit <- heatmap_fitness(df_moco)

# plot genomic locations
df_moco <- df_moco %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_moco_g1 <- genome_plot(df_moco, xlim = c(2455.9, 2458.6), title = "chr 1")
plot_moco_g2 <- genome_plot(df_moco, xlim = c(725.7, 728.3), title = "chr 1")

print(plot_moco_fit, position = c(0,0.23,1,1.05), more = TRUE)
print(plot_moco_g1, position = c(0.17,-0.1,0.57,0.4), more = TRUE)
print(plot_moco_g2, position = c(0.53,-0.1,0.83,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_moco.svg", width = 7, height = 4)
print(plot_moco_fit, position = c(0,0.23,1,1.05), more = TRUE)
print(plot_moco_g1, position = c(0.17,-0.1,0.57,0.4), more = TRUE)
print(plot_moco_g2, position = c(0.53,-0.1,0.83,0.4))
dev.off()
```

## Formate dehydrogenase

- fdsABCDG = primary soluble FDH
- fdwAB = secondary soluble FDH
- fdhABCD = primary membrane-bound FDH, 2 copies A1B1... and A2B2...
- fdoGHI = secondary membrane-bound FDH


```{r, fig.width = 7, fig.height = 4}
# plot heatmap
df_fdh <- df_fitness_summary %>% filter(time == 8, grepl("fd[swho]", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "fd[swho][A-Z0-9]+")) %>%
  mutate(gene_name = recode(gene_name, `fdhD` = "fdsC", `fdhC` = "fdhC1", `fdhE` = "fdhE1")) %>%
  ungroup %>% arrange(gene_name)
plot_fdh_fit <- heatmap_fitness(df_fdh)

# plot genomic locations
df_fdh <- df_fdh %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_fdh_g1 <- genome_plot(df_fdh, xlim = c(3167.9, 3174.9), title = "chr 1")
plot_fdh_g2 <- genome_plot(df_fdh, xlim = c(677.7, 685.6), title = "chr 1")

print(plot_fdh_fit, position = c(0,0.23,1,1.05), more = TRUE)
print(plot_fdh_g1, position = c(0.17,-0.1,0.56,0.4), more = TRUE)
print(plot_fdh_g2, position = c(0.52,-0.1,0.83,0.4))
```
```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_fdh.svg", width = 7, height = 4)
print(plot_fdh_fit, position = c(0,0.23,1,1.05), more = TRUE)
print(plot_fdh_g1, position = c(0.17,-0.1,0.56,0.4), more = TRUE)
print(plot_fdh_g2, position = c(0.52,-0.1,0.83,0.4))
dev.off()
```

## Hydrogenases

- hoxA = master regulator, 2-component system. HoxA is constantly active in its non-
  phosphorylated state below a temperature of 33*C. Originally, hoxA activity is regulated by
  the kinase HoxJ (together with HoxBC, the 'regulatory' hydrogenase). This function is inactivated
  due to a mutation in hoxJ,
  [Friedrich & Friedrich, J Bac, 1983](https://journals.asm.org/doi/10.1128/jb.153.1.176-181.1983),
  [Lenz et al., JMMB, 2004](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.334.5330&rep=rep1&type=pdf).
- hoxKGZ = membrane-bound hydrogenase
- hoxMLOQRTV = MBH accessory genes (cofactor incorporation, subunit assembly, 
  and twin-arginine-dependent membrane translocation - TAT),
  [Schubert et al., Mol Microbio, 2007](https://pubmed.ncbi.nlm.nih.gov/17850259/),
  [Fritsch et al., J Bac, 2011](https://pubmed.ncbi.nlm.nih.gov/21441514/)
  [Schwartz et al., J Mol Bio, 2003](https://pubmed.ncbi.nlm.nih.gov/12948488)
- hoxN1, hoxN2 = Nickel/Cobalt transporters. HoxN2 51% identical on AA level but not functional,
  [Schwartz et al., J Mol Bio, 2003](https://pubmed.ncbi.nlm.nih.gov/12948488)
- hoxBC = regulatory hydrogenase RH, oxygen-resistant non-energy generating hydrogen sensor. 
  HoxC is large active-site-containing subunit, hoxB small subunit HoxB carrying Fe-S clusters
- HoxJ =  RH binding signal transducer, histidine protein kinase
  [Löscher et al., Chem Phys Chem, 2010](https://pubmed.ncbi.nlm.nih.gov/20340124/)
  [Buerstel et al., PNAS, 2016](https://pubmed.ncbi.nlm.nih.gov/27930319/)
- hoxFUYH = soluble hydrogenase
- hoxW = soluble hydrogenase (HoxH-specific) C-terminal protease (maturation)
- hoxI = optional subunit of SH (hoxFUYH); specific reaction site for NADPH,
  [Burgdorf et al., J Bac, 2005](https://pubmed.ncbi.nlm.nih.gov/15838039/)


```{r, fig.width = 8, fig.height = 4}
# plot heatmap
df_hyd <- df_fitness_summary %>% filter(time == 8, grepl("hox", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "hox[A-Z0-9]+")) %>%
  ungroup %>% mutate(gene_name = factor(gene_name, paste0("hox", c("K","G","Z","M",
    "L","O","Q","R","T","V","A","B","C","J","N","N2","F","U","Y","H","W","I"))))
plot_hyd_fit <- heatmap_fitness(df_hyd)

# plot genomic locations
df_hyd <- df_hyd %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_hyd_g1 <- genome_plot(df_hyd, xlim = c(0, 9.0), title = "pHG1")
plot_hyd_g2 <- genome_plot(df_hyd, xlim = c(15.0, 22.9), title = "pHG1")
plot_hyd_g3 <- genome_plot(df_hyd, xlim = c(79.2, 85.8), title = "pHG1")

print(plot_hyd_fit, position = c(0,0.25,1,1), more = TRUE)
print(plot_hyd_g1, position = c(0.0,-0.1,0.39,0.4), more = TRUE)
print(plot_hyd_g2, position = c(0.36,-0.1,0.71,0.4), more = TRUE)
print(plot_hyd_g3, position = c(0.68,-0.1,1.0,0.4))
```
```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_hyd.svg", width = 8, height = 4)
print(plot_hyd_fit, position = c(0,0.25,1,1), more = TRUE)
print(plot_hyd_g1, position = c(0.0,-0.1,0.39,0.4), more = TRUE)
print(plot_hyd_g2, position = c(0.36,-0.1,0.71,0.4), more = TRUE)
print(plot_hyd_g3, position = c(0.68,-0.1,1.0,0.4))
dev.off()
```

## Hydrogenase pleiotropy genes (maturation)

- hypA1,B1,F1,C1,D1,E1 = Ni-Fe metallocenter formation proteins for both MBH and SH. "HypC1, hypD1, 
  and hypE1 are essential for SH and MBH maturation, while only one intact copy each of hypA,
  hypB, and hypF is needed", [Schwartz et al., J Mol Bio, 2003](https://pubmed.ncbi.nlm.nih.gov/12948488)
- hypX = CO cofactor formation for NiFe-hydrogenase. HypX converts formyl-THF into water and CO in 
  aerobic conditions, [Schulz et al., JACS, 2020](https://doi.org/10.1021/jacs.9b11506)
- hypA2,B2,F2 = alternative genes for hypA1,B1,F1, not sufficient alone for hydrogenase formation
- hypF3,hypC2,hypD2,hypE2,hypA3,hypB3. The physiological function of these extra hyp genes is unclear

```{r, fig.width = 7, fig.height = 4}
df_hyp <- df_fitness_summary %>% filter(time == 8, grepl("hyp", gene_name)) %>%
  # correct names of several hyp genes to make it more systematic
  mutate(gene_name = gsub("hypA PHG094", "hypA2 PHG094", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "hyp[A-Z0-9]+") %>%
    recode(hypA = "hypA1", hypB = "hypB1", hypC = "hypC1", hypD = "hypD1", hypE = "hypE1")) %>%
  ungroup %>% mutate(gene_name = factor(gene_name, paste0("hyp", c("A1","B1","C1","D1","E1",
    "F1","X","A2","B2","C2","D2","E2","F2","A3","B3","F3"))))
plot_hyp_fit <- heatmap_fitness(df_hyp)

# plot genomic locations
df_hyp <- df_hyp %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_hyp_g1 <- genome_plot(df_hyp, xlim = c(8.0, 16.0), title = "pHG1")
plot_hyp_g2 <- genome_plot(df_hyp, xlim = c(66.3, 73.5), title = "pHG1")
plot_hyp_g3 <- genome_plot(df_hyp, xlim = c(85.0, 89.7), title = "pHG1")

print(plot_hyp_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_hyp_g1, position = c(0.0,-0.1,0.39,0.4), more = TRUE)
print(plot_hyp_g2, position = c(0.36,-0.1,0.71,0.4), more = TRUE)
print(plot_hyp_g3, position = c(0.68,-0.1,1.0,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_hyp.svg", width = 7, height = 4)
print(plot_hyp_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_hyp_g1, position = c(0.0,-0.1,0.39,0.4), more = TRUE)
print(plot_hyp_g2, position = c(0.36,-0.1,0.71,0.4), more = TRUE)
print(plot_hyp_g3, position = c(0.68,-0.1,1.0,0.4))
dev.off()
```

## Other complexes important in energy metabolism

- **succinate dehydrogenase**
  - sdhABCD = membrane integral succinate dehydrogenase, reducing UQ to UQH2:
    **all of its genes are strictly essential**, no KO mutants available

- **NADH dehydrogenase**
  - nuoABCDEFGHIJKLMN = membrane integral NADH dehydrogenase, oxidizes NADH, pumps 4 H+,
    reduces UQ to UQH2: **all of its genes are strictly essential**, no KO mutants available

## Electron transport chain

Cytochrome e- transport proteins, cytochrome reductases, and terminal oxidases.

- **quinol oxidases**
  - cyoA1B1C1D1, ... (3 copies) = cytochrome bo3 complex, quinole oxidase, low O2 affinity
  - cydA1B1, ... (2 copies) = cytochrome bd complex, quinole oxidase, high O2 affinity
- **cytochrome reductases**
  - qcrABC (aka petABC, `H16_A3398`, `H16_A3397`, `H16_A3396`) = cytochrome bc1 reductase
- **cytochrome C e- carriers**
  - cytochrome c553 according to uniprot: `H16_A3451`, `H16_A0830`, `H16_A0846`,
    `H16_B1452`, `H16_A3576`, `H16_A1121`, `H16_A1120`. 
    According to String DB, `H16_A3576` is in same operon as thiosulphate oxidation
    pathway (SoxXYZABCD), where thiosulfate is used as electron source [Zhang et al., 2020](https://www.nature.com/articles/s41396-020-0684-5)
- **terminal cytochrome C oxidases**
  - coxMNOPQ = bb3 complex, affinity unknown?
  - ctaABCDEG = aa3 complex, low O2 affinity
  - ccoGNOP = cbb3 complex, high affinity, high similarity to NO3 reductase
  - VP2641 (`H16_A1031`) = Membrane spanning protein, RDD family. PFAM: unknown function, 
    potential transporter? Interacts with cytochrome c553 and qcrC, cytochrome reductase. 
    It's depletion implicates this protein in formate-specific e- transport in the membrane.

```{r, fig.width = 9, fig.height = 4}
df_etc <- df_fitness_summary %>% filter(time == 8, locus_tag %in% c("H16_A3451", "H16_A0830",
    "H16_A0846", "16_B1452", "H16_A3576", "H16_A1121", "H16_A1120", "H16_A1031") | 
    grepl("qcr|cco|cta|cox[MNOPQ] |cyo|cyd", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "H16_[AB][0-9]+|(qcr|cco|cta|cox|cyo|cyd)([A-Z0-9]+)?") %>%
    factor(c(
      paste0("cyo", c("A1","A2","A3","B1","B2","B3","C1","C2","D1","D2")),
      paste0("cyd", c("A1","B1","B2")), paste0("qcr", c("A","B","C")),
      paste0("H16_", c("A3576","A0846","A1120","A1121","A1031","A3451")),
      paste0("cox", c("M","N","O","P","Q")), paste0("cta", c("A","B","C","D","E","G")),
      paste0("cco", c("G","N","O","P"))
    ))
  )
plot_etc_fit <- heatmap_fitness(df_etc)

# plot genomic locations
df_etc <- df_etc %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
#plot_etc_g1 <- genome_plot(df_etc, xlim = c(1167.6, 1172.7), title = "chr 1")
plot_etc_g1 <- genome_plot(df_etc, xlim = c(3675.2, 3679.0), title = "chr 1")
plot_etc_g2 <- genome_plot(df_etc, xlim = c(361.2, 371.2), title = "chr 1")
plot_etc_g3 <- genome_plot(df_etc, xlim = c(2509.5, 2515.2), title = "chr 1")

print(plot_etc_fit, position = c(0,0.25,1,1), more = TRUE)
print(plot_etc_g1, position = c(0.1,-0.1,0.31,0.4), more = TRUE)
print(plot_etc_g2, position = c(0.28,-0.1,0.65,0.4), more = TRUE)
print(plot_etc_g3, position = c(0.62,-0.1,0.9,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_etc.svg", width = 9, height = 4)
print(plot_etc_fit, position = c(0,0.25,1,1), more = TRUE)
print(plot_etc_g1, position = c(0.1,-0.1,0.31,0.4), more = TRUE)
print(plot_etc_g2, position = c(0.28,-0.1,0.65,0.4), more = TRUE)
print(plot_etc_g3, position = c(0.62,-0.1,0.9,0.4))
dev.off()
```


## Nitrate respiration

**Nitrate reductase (NAR)**

Reaction: HNO3- + QH2 --> NO2- + H2O + Q.

Two copies of the *nar* operon, one on the megaplasmid and one on chromosome 2. Several other enzymes catalyze similar reactions as NAR. NAS is cytoplasmic and NADH dependent nitrate reductase [Cramm R, 2008](https://pubmed.ncbi.nlm.nih.gov/18957861/). NAP is membrane protein complex accepting QH2, involved in nitrate respiration but probably also dissipation of excess redox power. Here we just focus on the main NAR enzyme.

```{r, fig.width = 6, fig.height = 4}
df_nar <- df_fitness_summary %>%
  filter(time == 8, grepl("nar", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "nar([A-Z0-9]+)?") %>%
    recode(narG = "narG1", narH = "narH1", narI = "narI1", narL = "narL1", narX = "narX1") %>%
    factor(., paste0("nar", c("G1","H1","I1","K1","L1","X1","G2","H2","I2","K2","L2","X2","K3","K5"))))
plot_nar_fit <- heatmap_fitness(df_nar)

# plot genomic locations
df_nar <- df_nar %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_nar_g1 <- genome_plot(df_nar, xlim = c(279.5, 294.5), title = "pHG1")

print(plot_nar_fit, position = c(0,0.25,1,1), more = TRUE)
print(plot_nar_g1, position = c(0.2,-0.1,0.85,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_nar.svg", width = 6, height = 4)
print(plot_nar_fit, position = c(0,0.25,1,1), more = TRUE)
print(plot_nar_g1, position = c(0.2,-0.1,0.85,0.4))
dev.off()
```

**Nitrite reductase (NIR)**

Reaction: NO2- + QH2 --> NO + H2O + Q

NIR is a cd1 type copper containing terminal oxidase, homodimer of 2 nirS subunits.
The genes for nitrite reductase are assembled in a cluster  (nirS, -C, -F, -D, -G, -H, -J, -N, -E)   spanning 8.6  kb  on  chromosome  2. The  first  gene  of  this  cluster,  nirS,  is the structural gene encoding  cd1  -NIR.  The  re-maining genes are presumed to be accessory genes essential  for  maturation  of  active  nitrite  reductase (?).

```{r, fig.width = 6, fig.height = 4}
df_nir <- df_fitness_summary %>%
  filter(time == 8, grepl("nir", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "nir([A-Z0-9]+)?"))
plot_nir_fit <- heatmap_fitness(df_nir)

# plot genomic locations
df_nir <- df_nir %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_nir_g1 <- genome_plot(df_nir, xlim = c(2576.3, 2585.9), title = "chr 2")

print(plot_nir_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_nir_g1, position = c(0.3,-0.1,0.75,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_nir.svg", width = 6, height = 4)
print(plot_nir_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_nir_g1, position = c(0.3,-0.1,0.75,0.4))
dev.off()
```

**NO reductase (NOR)**

Reaction: NO + QH2 --> N2O + H2O + Q

From [Cramm R, 2008](https://pubmed.ncbi.nlm.nih.gov/18957861/): "The catalytic sub-unit (NorB) of all NORs that have been characterized to date contains one high-spin heme b and a binuclear catalytic center of a low-spin heme b and a non-heme iron. The best investigated NORs are heterodimers that contain, in addition to NorB, a heme c-containing subunit NorC which channels lectrons from the physiological electron donor cytochrome c to NorB."

```{r, fig.width = 6, fig.height = 4}
df_nor <- df_fitness_summary %>%
  filter(time == 8, grepl("nor", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "nor([A-Z0-9]+)?"))
plot_nor_fit <- heatmap_fitness(df_nor)

# plot genomic locations
df_nor <- df_nor %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_nor_g1 <- genome_plot(df_nor, xlim = c(256.4, 262.1), title = "pHG1")

print(plot_nor_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_nor_g1, position = c(0.3,-0.1,0.75,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_nor.svg", width = 6, height = 4)
print(plot_nor_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_nor_g1, position = c(0.3,-0.1,0.75,0.4))
dev.off()
```

**NOS (nitrous oxide) reductase**

Reaction: N2O + QH2 --> N2 + H2O + Q

The NOS of R. eutropha H16 has not been investigated on the biochemical level. Genes for NOS are located adjacent to the nor genes on pHG1. NosC (PHG253) might be the cyt c that is the e- donor for NOS. The nosC KO is enriched on formate, i.e. grows better than WT. Is nosC taking e- away from formate-fed ETC?

```{r, fig.width = 6, fig.height = 4}
df_nos <- df_fitness_summary %>%
  filter(time == 8, grepl("nos", gene_name)) %>%
  mutate(gene_name = stri_extract_first_regex(gene_name, "nos([A-Z0-9]+)?"))
plot_nos_fit <- heatmap_fitness(df_nos)

# plot genomic locations
df_nos <- df_nos %>% group_by(locus_tag, gene_name, scaffold, strand, start, end) %>%
  summarize(strains_per_gene = min(unique(strains_per_gene)), .groups = "drop")
plot_nos_g1 <- genome_plot(df_nos, xlim = c(261.1, 271.9), title = "pHG1")

print(plot_nos_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_nos_g1, position = c(0.22,-0.1,0.85,0.4))
```

```{r, echo = FALSE, results = 'hide'}
svg(filename = "../figures/figure_fit_nos.svg", width = 6, height = 4)
print(plot_nos_fit, position = c(0,0.25,1,1.0), more = TRUE)
print(plot_nos_g1, position = c(0.22,-0.1,0.85,0.4))
dev.off()
```

## Main conclusions

- MoCo cofactor is important for growth on formate *and* nitrate respiration, but not for growth on hydrogen
- Main FDH is S-FDH both in terms of protein abundance, inducibility, and essentiality
- Alternative S-FDH operon *fdw* does not play a roll
- Both hydrogenases, S-H and m_H have similar fitness penalty on hydrogen when KOed
- Hydrogenase genes, but even more so master regulator hoxA and Hydrogenase pleiotropy (hyp) genes confer fitness advantage when KOed on all non-loithoautotrophic substrates
- Cytochrome reductase bc1 complex (qcrABC) is particularly important for growth on formate
- Bc1 complex plays no role on nitrate, therefore no cross talk (while a bc1 KO mutant was growth-deficient in oxygen-limited growth on nitrate (Garg et al., 2018).
- Bc1 complex accepts reduced quinols as e- donor, pumps 4 H+ /2 e− via proton-motive Q cycle (quinol:cyt c oxidoreduction), and releases reduced cytochrome C e- carriers. Main cytochrome complex for most substrates including hydrogen
- Cbb3-type cytochrome oxidase (*cco* operon) also important for growth on formate, while *cta* operon (aa3 complex) more important on fructose and succinate
 
# Session Info

```{r}
sessionInfo()
```